File: go/stream.go
package buble

import (
	"bufio"
	"encoding/json"
	"io"
	"strings"
)

// StreamProtocol identifies the protocol shape used by an SSE stream.
type StreamProtocol string

const (
	// StreamProtocolOpenAI extracts text from OpenAI-compatible chunks.
	StreamProtocolOpenAI StreamProtocol = "openai"
	// StreamProtocolAnthropic extracts text from Anthropic-compatible events.
	StreamProtocolAnthropic StreamProtocol = "anthropic"
	// StreamProtocolGemini extracts text from Gemini-compatible chunks.
	StreamProtocolGemini StreamProtocol = "gemini"
)

// Stream reads server-sent events from a streaming chat response.
type Stream struct {
	reader   io.ReadCloser
	scanner  *bufio.Scanner
	protocol StreamProtocol
	event    SSEEvent
	err      error
	closed   bool
	buffer   []string
}

// NewStream creates a Stream from a response body and protocol.
func NewStream(reader io.ReadCloser, protocol StreamProtocol) *Stream {
	scanner := bufio.NewScanner(reader)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	return &Stream{
		reader:   reader,
		scanner:  scanner,
		protocol: protocol,
	}
}

// Next advances the stream to the next event. It returns false when the stream
// ends, the server sends [DONE], or an error occurs.
func (s *Stream) Next() bool {
	for s.scanner.Scan() {
		line := strings.TrimSuffix(s.scanner.Text(), "\r")
		if line == "" {
			if event, ok := parseSSEBlock(s.buffer); ok {
				s.buffer = nil
				if event.Data == "[DONE]" {
					_ = s.Close()
					return false
				}
				s.event = event
				return true
			}
			s.buffer = nil
			continue
		}
		s.buffer = append(s.buffer, line)
	}
	if err := s.scanner.Err(); err != nil {
		s.err = err
		_ = s.Close()
		return false
	}
	if event, ok := parseSSEBlock(s.buffer); ok {
		s.buffer = nil
		if event.Data == "[DONE]" {
			_ = s.Close()
			return false
		}
		s.event = event
		return true
	}
	_ = s.Close()
	return false
}

// Event returns the current SSE event after a successful call to Next.
func (s *Stream) Event() SSEEvent {
	return s.event
}

// Text returns the text delta from the current event, when available.
func (s *Stream) Text() string {
	return textFromEvent(s.event, s.protocol)
}

// Err returns the stream read error, if any.
func (s *Stream) Err() error {
	return s.err
}

// Close closes the underlying response body.
func (s *Stream) Close() error {
	if s.closed {
		return nil
	}
	s.closed = true
	return s.reader.Close()
}

func parseSSEBlock(lines []string) (SSEEvent, bool) {
	var event SSEEvent
	var dataLines []string
	for _, raw := range lines {
		line := strings.TrimRight(raw, "\r")
		if line == "" || strings.HasPrefix(line, ":") {
			continue
		}
		field, value, found := strings.Cut(line, ":")
		if found {
			value = strings.TrimPrefix(value, " ")
		}
		switch field {
		case "event":
			event.Event = value
		case "data":
			dataLines = append(dataLines, value)
		}
	}
	if event.Event == "" && len(dataLines) == 0 {
		return SSEEvent{}, false
	}
	event.Data = strings.Join(dataLines, "\n")
	if event.Data != "" && event.Data != "[DONE]" {
		var parsed any
		if err := json.Unmarshal([]byte(event.Data), &parsed); err == nil {
			event.JSON = parsed
		}
	}
	return event, true
}

func textFromEvent(event SSEEvent, protocol StreamProtocol) string {
	payload, ok := event.JSON.(map[string]any)
	if !ok {
		return ""
	}
	switch protocol {
	case StreamProtocolOpenAI:
		return textFromOpenAI(payload)
	case StreamProtocolAnthropic:
		if event.Event != "content_block_delta" {
			return ""
		}
		return textFromAnthropic(payload)
	case StreamProtocolGemini:
		return textFromGemini(payload)
	default:
		return ""
	}
}

func textFromOpenAI(payload map[string]any) string {
	choices, ok := payload["choices"].([]any)
	if !ok || len(choices) == 0 {
		return ""
	}
	choice, ok := choices[0].(map[string]any)
	if !ok {
		return ""
	}
	delta, ok := choice["delta"].(map[string]any)
	if !ok {
		return ""
	}
	text, _ := delta["content"].(string)
	return text
}

func textFromAnthropic(payload map[string]any) string {
	delta, ok := payload["delta"].(map[string]any)
	if !ok {
		return ""
	}
	text, _ := delta["text"].(string)
	return text
}

func textFromGemini(payload map[string]any) string {
	choices, ok := payload["candidates"].([]any)
	if !ok || len(choices) == 0 {
		return ""
	}
	choice, ok := choices[0].(map[string]any)
	if !ok {
		return ""
	}
	content, ok := choice["content"].(map[string]any)
	if !ok {
		return ""
	}
	parts, ok := content["parts"].([]any)
	if !ok {
		return ""
	}
	var out strings.Builder
	for _, partValue := range parts {
		part, ok := partValue.(map[string]any)
		if !ok {
			continue
		}
		text, ok := part["text"].(string)
		if ok {
			out.WriteString(text)
		}
	}
	return out.String()
}

File: npm/src/stream.ts
import type { SSEEvent } from './types/chat.js';

function tryParseJson(data: string) {
  if (!data || data === '[DONE]') return undefined;
  try {
    return JSON.parse(data);
  } catch {
    return undefined;
  }
}

function parseSSEBlock(block: string): SSEEvent | undefined {
  let event: string | undefined;
  const dataLines: string[] = [];

  for (const rawLine of block.split(/\r?\n/)) {
    const line = rawLine.trimEnd();
    if (!line || line.startsWith(':')) continue;
    const separator = line.indexOf(':');
    const field = separator === -1 ? line : line.slice(0, separator);
    const value = separator === -1 ? '' : line.slice(separator + 1).replace(/^ /, '');

    if (field === 'event') event = value;
    if (field === 'data') dataLines.push(value);
  }

  if (!event && dataLines.length === 0) return undefined;

  const data = dataLines.join('\n');
  return { event, data, json: tryParseJson(data) };
}

export async function* parseSSEStream(
  body: ReadableStream<Uint8Array> | null
): AsyncIterable<SSEEvent> {
  if (!body) return;

  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      while (true) {
        const match = buffer.match(/\r?\n\r?\n/);
        if (!match?.index && match?.index !== 0) break;

        const block = buffer.slice(0, match.index);
        buffer = buffer.slice(match.index + match[0].length);
        const event = parseSSEBlock(block);
        if (event) yield event;
      }
    }

    buffer += decoder.decode();
    const event = parseSSEBlock(buffer);
    if (event) yield event;
  } finally {
    reader.releaseLock();
  }
}

function textFromOpenAIChunk(value: unknown) {
  const chunk = value as any;
  return chunk?.choices?.[0]?.delta?.content;
}

function textFromAnthropicEvent(value: SSEEvent) {
  const json = value.json as any;
  if (value.event === 'content_block_delta') {
    return json?.delta?.text;
  }
  return undefined;
}

function textFromGeminiChunk(value: unknown) {
  const chunk = value as any;
  return chunk?.candidates?.[0]?.content?.parts
    ?.map((part: any) => part?.text)
    .filter(Boolean)
    .join('');
}

export async function* streamText(
  events: AsyncIterable<SSEEvent>,
  protocol: 'openai' | 'anthropic' | 'gemini'
): AsyncIterable<string> {
  for await (const event of events) {
    if (event.data === '[DONE]') break;
    const text =
      protocol === 'openai'
        ? textFromOpenAIChunk(event.json)
        : protocol === 'anthropic'
          ? textFromAnthropicEvent(event)
          : textFromGeminiChunk(event.json);
    if (typeof text === 'string' && text.length > 0) {
      yield text;
    }
  }
}

export class BubleStream implements AsyncIterable<SSEEvent> {
  readonly events: AsyncIterable<SSEEvent>;
  readonly protocol: 'openai' | 'anthropic' | 'gemini';

  constructor(
    events: AsyncIterable<SSEEvent>,
    protocol: 'openai' | 'anthropic' | 'gemini'
  ) {
    this.events = events;
    this.protocol = protocol;
  }

  [Symbol.asyncIterator]() {
    return this.events[Symbol.asyncIterator]();
  }

  toTextStream() {
    return streamText(this.events, this.protocol);
  }
}
